File: java/DFS.java
import java.util.*;

public class Main {
    // Function to add an edge to the adjacency list
    public static void
    addEdge(ArrayList<ArrayList<Integer> > adj, int s,
            int t)
    {
        adj.get(s).add(t);
        adj.get(t).add(s);
    }

    // Recursive function for DFS traversal
    private static void
    dfsRec(ArrayList<ArrayList<Integer> > adj,
           boolean[] visited, int s, ArrayList<Integer> res)
    {
        visited[s] = true;
        res.add(s);

        // Recursively visit all adjacent vertices that are
        // not visited yet
        for (int i : adj.get(s)) {
            if (!visited[i]) {
                dfsRec(adj, visited, i, res);
            }
        }
    }

    // Main DFS function to perform DFS for the entire graph
    public static ArrayList<Integer>
    DFS(ArrayList<ArrayList<Integer> > adj)
    {
        boolean[] visited = new boolean[adj.size()];
        ArrayList<Integer> res = new ArrayList<>();

        // Loop through all vertices to handle disconnected
        // graphs
        for (int i = 0; i < adj.size(); i++) {
            if (!visited[i]) {
                dfsRec(adj, visited, i, res);
            }
        }

        return res;
    }

    public static void main(String[] args)
    {
        int V = 6;
        // Create an adjacency list for the graph
        ArrayList<ArrayList<Integer> > adj
            = new ArrayList<>();

        // Initialize adjacency list
        for (int i = 0; i < V; i++) {
            adj.add(new ArrayList<>());
        }

        // Define the edges of the graph
        int[][] edges
            = { { 1, 2 }, { 2, 0 }, { 0, 3 }, { 4, 5 } };

        // Populate the adjacency list with edges
        for (int[] e : edges) {
            addEdge(adj, e[0], e[1]);
        }

        // Perform DFS
        ArrayList<Integer> res = DFS(adj);

        // Print the DFS traversal result
        for (int num : res) {
            System.out.print(num + " ");
        }
    }
}


// Output
// 0 2 1 3 4 5 

// Time complexity: O(V + E). Note that the time complexity is same here because we visit every vertex at most once and every edge is traversed at most once (in directed) and twice in undirected.
// Auxiliary Space: O(V + E), since an extra visited array of size V is required, And stack size for recursive calls to dfsRec function.

File: javascript/dfs.js
// Graph class using adjacency list
class Graph {
    constructor() {
        this.adjList = new Map();
    }

    // Add a vertex
    addVertex(vertex) {
        if (!this.adjList.has(vertex)) {
            this.adjList.set(vertex, []);
        }
    }

    // Add an edge (undirected graph)
    addEdge(v, w) {
        if (this.adjList.has(v)) this.adjList.get(v).push(w);
        if (this.adjList.has(w)) this.adjList.get(w).push(v);
    }

    // Depth-First Search
    dfs(startingVertex) {
        const visited = new Set();
        const result = [];

        const dfsRecursive = (vertex) => {
            visited.add(vertex);
            result.push(vertex);

            const neighbors = this.adjList.get(vertex);
            for (const neighbor of neighbors) {
                if (!visited.has(neighbor)) {
                    dfsRecursive(neighbor);
                }
            }
        };

        dfsRecursive(startingVertex);
        return result;
    }
}

// Example usage
const g = new Graph();

// Add vertices
['A', 'B', 'C', 'D', 'E', 'F'].forEach(v => g.addVertex(v));

// Add edges
g.addEdge('A', 'B');
g.addEdge('A', 'C');
g.addEdge('B', 'D');
g.addEdge('B', 'E');
g.addEdge('C', 'F');

// Perform DFS starting from vertex 'A'
const dfsResult = g.dfs('A');
console.log("DFS Traversal Order:", dfsResult.join(' -> '));
// Output: DFS Traversal Order: A -> B -> D -> E -> C -> F
